pkg/reconciler/certificate/certificate.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificate
+
+import (
+	context "context"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"knative.dev/net-http01/pkg/ordermanager"
+	"knative.dev/net-http01/pkg/reconciler/certificate/resources"
+	logging "knative.dev/pkg/logging"
+	reconciler "knative.dev/pkg/reconciler"
+	v1alpha1 "knative.dev/serving/pkg/apis/networking/v1alpha1"
+	certificate "knative.dev/serving/pkg/client/injection/reconciler/networking/v1alpha1/certificate"
+)
+
+// Reconciler implements controller.Reconciler for Certificate resources.
+type Reconciler struct {
+	kubeClient kubernetes.Interface
+
+	secretLister    corev1listers.SecretLister
+	serviceLister   corev1listers.ServiceLister
+	endpointsLister corev1listers.EndpointsLister
+
+	orderManager ordermanager.Interface
+}
+
+// Check that our Reconciler implements Interface
+var _ certificate.Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *v1alpha1.Certificate) reconciler.Event {
+	o.Status.InitializeConditions()
+
+	svc, err := r.reconcileService(ctx, o)
+	if err != nil {
+		return err
+	}
+	if err := r.reconcileEndpoints(ctx, o); err != nil {
+		return err
+	}
+
+	// Lookup the secret, and ensure that it's contents are still valid.
+	secret, err := r.secretLister.Secrets(o.Namespace).Get(o.Spec.SecretName)
+	if apierrs.IsNotFound(err) {
+		// We have to create it!
+		logging.FromContext(ctx).Info("Secret doesn't exist, we must provision a new Certificate.")
+	} else if err != nil {
+		return err
+	} else if valid, err := resources.IsValidCertificate(secret, o.Spec.DNSNames, 48*time.Hour); err == nil && valid {
```

Comment:
we probably want to attempt renewal 30 days before the expiration https://docs.https.dev/acme-ops#renew-certificates-after-of-usable-lifetime

Reply:
done